File: server/test/fazoura_web/channels/protocol_scenarios_test.exs
defmodule FazouraWeb.ProtocolScenariosTest do
  @moduledoc """
  Replays every `protocol/fixtures/scenarios/*.json` script against the Phoenix
  implementation (PROTOCOL.md §11). The LAN host must pass the same scripts.

  Each actor is its own process acting as a socket transport, so pushes and replies
  are attributed to the right client.
  """

  use FazouraWeb.ChannelCase, async: true

  alias Fazoura.Game.Pack
  alias Fazoura.ProtocolFixtures
  alias Fazoura.Rooms

  @clock_start 1_789_502_400_000
  @wait_ms 1_000

  scenarios = Path.wildcard(Path.join(ProtocolFixtures.dir(), "scenarios/*.json"))
  if scenarios == [], do: raise("no protocol scenarios found in #{ProtocolFixtures.dir()}")

  for path <- scenarios do
    @external_resource path
    scenario = path |> File.read!() |> Jason.decode!()

    test "scenario #{Path.basename(path)}: #{scenario["name"]}" do
      run_scenario(unquote(Macro.escape(scenario)))
    end
  end

  ## Runner

  defp run_scenario(scenario) do
    {:ok, clock} = Agent.start_link(fn -> @clock_start end)
    pack = Pack.from_map(scenario["pack"])
    {:ok, code, host_token} = Rooms.create(pack, now: fn -> Agent.get(clock, & &1) end)

    ctx = %{code: code, clock: clock, actors: %{}, vars: %{"$host_token" => host_token}}

    ctx =
      scenario["steps"]
      |> Enum.with_index(1)
      |> Enum.reduce(ctx, fn {step, index}, ctx ->
        run_step(step, ctx, "step #{index}: #{Jason.encode!(step)}")
      end)

    # Stop the room while the fake clock it reads is still alive.
    with [{pid, _}] <- Registry.lookup(Fazoura.Rooms.Registry, ctx.code) do
      DynamicSupervisor.terminate_child(Fazoura.Rooms.Supervisor, pid)
    end
  end

  defp run_step(%{"advance_clock_ms" => ms}, ctx, _label) do
    Agent.update(ctx.clock, &(&1 + ms))
    :ok = Rooms.tick(ctx.code)
    ctx
  end

  defp run_step(%{"actor" => name, "join" => payload} = step, ctx, label) do
    actor = start_actor(socket(FazouraWeb.UserSocket, nil, %{}))
    reply = actor_call(actor, {:join, "room:" <> ctx.code, subst(payload, ctx)})
    assert_matches(subst(step["expect"], ctx), reply, label)

    vars =
      case reply do
        %{"status" => "ok", "response" => %{"player_id" => id, "player_token" => token}}
        when is_binary(id) ->
          Map.merge(ctx.vars, %{"$player_id:#{name}" => id, "$player_token:#{name}" => token})

        _ ->
          ctx.vars
      end

    if reply["status"] == "ok",
      do: %{ctx | actors: Map.put(ctx.actors, name, actor), vars: vars},
      else: ctx
  end

  defp run_step(%{"actor" => name, "push" => event} = step, ctx, label) do
    reply = actor_call(actor!(ctx, name), {:push, event, subst(step["payload"], ctx)})
    assert_matches(subst(step["expect"], ctx), reply, label)
    ctx
  end

  defp run_step(%{"actor" => name, "expect_state" => expected}, ctx, label) do
    wait_for_state(
      actor!(ctx, name),
      subst(expected, ctx),
      label,
      System.monotonic_time(:millisecond)
    )

    ctx
  end

  defp run_step(%{"actor" => name, "disconnect" => true}, ctx, _label) do
    actor = actor!(ctx, name)
    ref = Process.monitor(actor)
    Process.exit(actor, :kill)
    assert_receive {:DOWN, ^ref, _, _, _}
    %{ctx | actors: Map.delete(ctx.actors, name)}
  end

  defp actor!(ctx, name), do: Map.get(ctx.actors, name) || flunk("unknown actor #{name}")

  defp wait_for_state(actor, expected, label, started) do
    state = actor_call(actor, :latest_state)

    cond do
      matches?(expected, state) ->
        :ok

      System.monotonic_time(:millisecond) - started > @wait_ms ->
        flunk(
          "#{label}\nexpected state to match:\n#{inspect(expected)}\nlatest state:\n#{inspect(state)}"
        )

      true ->
        Process.sleep(10)
        wait_for_state(actor, expected, label, started)
    end
  end

  ## Placeholders and partial matching (PROTOCOL.md §11)

  defp subst(value, ctx) when is_binary(value), do: Map.get(ctx.vars, value, value)

  defp subst(value, ctx) when is_map(value),
    do: Map.new(value, fn {k, v} -> {k, subst(v, ctx)} end)

  defp subst(value, ctx) when is_list(value), do: Enum.map(value, &subst(&1, ctx))
  defp subst(value, _ctx), do: value

  defp assert_matches(expected, actual, label) do
    assert matches?(expected, actual),
           "#{label}\nexpected:\n#{inspect(expected)}\ngot:\n#{inspect(actual)}"
  end

  defp matches?(expected, actual) when is_map(expected) and is_map(actual),
    do: Enum.all?(expected, fn {k, v} -> Map.has_key?(actual, k) and matches?(v, actual[k]) end)

  defp matches?(expected, actual) when is_list(expected) and is_list(actual),
    do:
      length(expected) == length(actual) and
        Enum.all?(Enum.zip(expected, actual), fn {e, a} -> matches?(e, a) end)

  defp matches?(expected, actual), do: expected == actual

  ## Actor process: one per client, acting as that client's transport

  defp start_actor(socket) do
    actor = spawn(fn -> actor_loop(nil, nil) end)
    send(actor, {:socket, %{socket | transport_pid: actor}})
    actor
  end

  defp actor_call(actor, request) do
    ref = make_ref()
    send(actor, {:call, self(), ref, request})

    receive do
      {^ref, result} -> result
    after
      5_000 -> flunk("actor did not answer #{inspect(request)}")
    end
  end

  defp actor_loop(socket, latest_state) do
    receive do
      {:socket, socket} ->
        actor_loop(socket, latest_state)

      %Phoenix.Socket.Message{event: "state", payload: payload} ->
        actor_loop(socket, wire(payload))

      %Phoenix.Socket.Message{} ->
        actor_loop(socket, latest_state)

      {:call, from, ref, {:join, topic, payload}} ->
        case join(socket, FazouraWeb.RoomChannel, topic, payload) do
          {:ok, reply, joined} ->
            send(from, {ref, wire(%{status: "ok", response: reply})})
            actor_loop(joined, latest_state)

          {:error, reply} ->
            send(from, {ref, wire(%{status: "error", response: reply})})
            actor_loop(socket, latest_state)
        end

      {:call, from, ref, {:push, event, payload}} ->
        push_ref = push(socket, event, payload)

        receive do
          %Phoenix.Socket.Reply{ref: ^push_ref, status: status, payload: reply} ->
            send(from, {ref, wire(%{status: status, response: reply})})
        after
          5_000 -> send(from, {ref, :no_reply})
        end

        actor_loop(socket, latest_state)

      {:call, from, ref, :latest_state} ->
        send(from, {ref, latest_state})
        actor_loop(socket, latest_state)
    end
  end

  # Round-trip through JSON so assertions see exactly what goes over the wire.
  defp wire(term), do: term |> Jason.encode!() |> Jason.decode!()
end

File: app/test/core/game/protocol_scenarios_test.dart
/// Replays every `protocol/fixtures/scenarios/*.json` script against the LAN
/// host, exactly as `FazouraWeb.ProtocolScenariosTest` replays them against
/// Phoenix (PROTOCOL.md §11).
///
/// This is what holds the Dart port of the game rules to the Elixir one. The
/// two implementations share no code and never will — a LAN host has no
/// Elixir — so the fixtures are the only thing that can catch them drifting
/// apart (AGENTS.md §3, §8).
///
/// The scripts are replayed against [LanRoom] rather than through a socket:
/// the room is the whole of the contract's behaviour, and a fake connection
/// per actor gives the per-recipient snapshots the scenarios assert on.
/// `lan_host_test.dart` covers the socket layer above it.
@TestOn('vm')
library;

import 'dart:convert';

import 'package:fazoura_party/core/game/game.dart';
import 'package:fazoura_party/core/game/lan_room.dart';
import 'package:fazoura_party/core/game/pack.dart';
import 'package:flutter_test/flutter_test.dart';

import '../../support/protocol_fixtures.dart';

/// The clock every scenario starts at, matching the Elixir runner — the
/// fixtures assert on absolute deadlines computed from it.
const _clockStart = 1789502400000;

void main() {
  for (final entry in ProtocolFixtures.scenarios()) {
    test('scenario ${entry.name}: ${entry.scenario['name']}', () {
      _ScenarioRun(entry.scenario).run();
    });
  }
}

class _ScenarioRun {
  _ScenarioRun(this.scenario);

  final Map<String, dynamic> scenario;
  final Map<String, _FakeConnection> actors = {};
  final Map<String, Object?> vars = {};

  int clock = _clockStart;
  late final LanRoom room = LanRoom.create(
    pack: Pack.fromMap(scenario['pack'] as Map<String, dynamic>),
    now: () => clock,
    // The scripts name the question they expect next, so the pack has to be
    // played in its written order — `shuffle_questions?: false` on the server.
    shuffleQuestions: false,
  );

  void run() {
    vars[r'$host_token'] = room.hostToken;
    try {
      final steps = scenario['steps'] as List;
      for (final (index, step) in steps.indexed) {
        _step(
          step as Map<String, dynamic>,
          'step ${index + 1}: ${jsonEncode(step)}',
        );
      }
    } finally {
      room.close(LanCloseReason.shutdown);
    }
  }

  void _step(Map<String, dynamic> step, String label) {
    if (step['advance_clock_ms'] case final int ms) {
      clock += ms;
      // The room's own timer cannot fire against an injected clock, so the
      // runner advances it by hand — what `Rooms.tick/1` is for on the server.
      room.tick();
      return;
    }

    final name = step['actor'] as String;

    if (step['join'] != null) {
      _join(name, step, label);
      return;
    }
    if (step['push'] case final String event) {
      final reply = _reply(() {
        room.handle(
          _actor(name, label),
          event,
          _substitute(step['payload']) as Map<String, dynamic>? ?? {},
        );
        return null;
      });
      _expectMatches(_substitute(step['expect']), reply, label);
      return;
    }
    if (step['expect_state'] != null) {
      _expectMatches(
        _substitute(step['expect_state']),
        _actor(name, label).latestState,
        label,
      );
      return;
    }
    if (step['expect_closed'] case final String reason) {
      expect(_actor(name, label).closedReason, reason, reason: label);
      return;
    }
    if (step['disconnect'] == true) {
      room.leave(_actor(name, label));
      actors.remove(name);
      return;
    }
    fail('$label: unknown step');
  }

  void _join(String name, Map<String, dynamic> step, String label) {
    final connection = _FakeConnection();
    LanJoinReply? joined;

    final reply = _reply(() {
      joined = room.join(
        connection,
        _substitute(step['join']) as Map<String, dynamic>,
      );
      return {
        'role': joined!.role,
        'player_id': joined!.playerId,
        'player_token': joined!.playerToken,
      };
    });

    _expectMatches(_substitute(step['expect']), reply, label);

    final result = joined;
    if (result == null) return;
    actors[name] = connection;
    if (result.playerId != null) {
      vars[r'$player_id:' + name] = result.playerId;
      vars[r'$player_token:' + name] = result.playerToken;
    }
  }

  /// Runs an intent and shapes the outcome like a `phx_reply` payload, which
  /// is what the fixtures' `expect` blocks describe (§4.2).
  Map<String, dynamic> _reply(Map<String, dynamic>? Function() intent) {
    try {
      return _wire({'status': 'ok', 'response': intent() ?? {}});
    } on GameRuleError catch (error) {
      return _wire({
        'status': 'error',
        'response': {'code': error.code, 'message': 'ignored'},
      });
    }
  }

  _FakeConnection _actor(String name, String label) =>
      actors[name] ?? (fail('$label: unknown actor $name'));

  /// Substitutes the `$host_token`, `$player_id:<actor>` and
  /// `$player_token:<actor>` placeholders (§11).
  Object? _substitute(Object? value) => switch (value) {
    String() => vars.containsKey(value) ? vars[value] : value,
    Map() => {
      for (final entry in value.entries)
        entry.key as String: _substitute(entry.value),
    },
    List() => [for (final item in value) _substitute(item)],
    _ => value,
  };

  void _expectMatches(Object? expected, Object? actual, String label) {
    expect(
      _matches(expected, actual),
      isTrue,
      reason: '$label\nexpected:\n$expected\ngot:\n$actual',
    );
  }

  /// Partial match, the same rule the server's runner uses (§11): every key
  /// present in `expected` must be present and equal, keys it omits are
  /// unchecked, and lists must be the same length element for element.
  static bool _matches(Object? expected, Object? actual) {
    if (expected is Map && actual is Map) {
      return expected.entries.every(
        (entry) =>
            actual.containsKey(entry.key) &&
            _matches(entry.value, actual[entry.key]),
      );
    }
    if (expected is List && actual is List) {
      return expected.length == actual.length &&
          List.generate(
            expected.length,
            (i) => _matches(expected[i], actual[i]),
          ).every((matched) => matched);
    }
    return expected == actual;
  }
}

/// Round-trips through JSON so assertions see exactly what goes over the wire,
/// and so a value the snapshot cannot encode fails here rather than on a
/// socket.
Map<String, dynamic> _wire(Map<String, dynamic> value) =>
    jsonDecode(jsonEncode(value)) as Map<String, dynamic>;

/// One client, from the room's point of view.
class _FakeConnection implements LanConnection {
  Map<String, dynamic>? latestState;
  String? closedReason;

  @override
  void pushState(Map<String, dynamic> state) => latestState = _wire(state);

  @override
  void pushClosed(String reason) => closedReason = reason;
}
